### PR TITLE
[GO] GetParam Endpoint

### LINKED ISSUE #1368: [GO] Get Param Endpoint
## 📖 Description
```
get:
  summary: Retrieves a parameter descriptor from the device model.
  operationId: Param

  # default security is read-only

  parameters:
    - name: slot
      in: path
      required: true
      description: The slot of the device model containing the parameter to retrieve.
      schema:
        $ref: "../../schemata/device.json#/$defs/slot"

    - name: fqoid
      in: path
      required: true
      description: The fully qualified path to the parameter within the device model.
      schema:
        $ref: "../../schemata/device.json#/$defs/fqoid"
      example: "text_box"

  tags:
    - Parameters

  responses:
    "200":
      description: The parameter descriptor specified by `fqoid` within the device model.
      content:
        application/json:
          schema:
            $ref: "../../schemata/device.json#/$defs/component_param"
          example:
            {
              "oid": "text_box",
              "param": {
                "name": {
                  "display_strings": {
                    "en": "Text Box",
                    "es": "Caja de Texto",
                    "fr": "Boîte de Texte"
                  }
                },
                "type": "STRING",
                "value": {
                  "string_value": "Hello, World!"
                }
              }
            }

    "400":
      $ref: '../openapi.yaml#/components/responses/BadRequest'

    "401":
      $ref: '../openapi.yaml#/components/responses/Unauthorized'

    "403":
      $ref: '../openapi.yaml#/components/responses/Forbidden'

    "404":
      $ref: '../openapi.yaml#/components/responses/NotFound'

    "500":
      $ref: '../openapi.yaml#/components/responses/InternalServerError'

    "503":
      $ref: '../openapi.yaml#/components/responses/ServiceUnavailable'
```
---

## 🎯 Goal / Outcome
url/slot/param/fqoid request returns appropriate param information
unit tests

---

## ✅ Definition of Done (DoD)

---

## 🛠️ Implementation Notes (Optional)

---

## 🔗 Related Issues / Links

---
## 🚧 Risks / Open Questions (Optional)

---

### LINKED ISSUE #1378: [GO] GRPC GetParam
## 📖 Description
Implement GetParam endpoint
```
message GetParamPayload {
  uint32 slot = 1; // Uniquely identifies the device at node scope.
  string oid = 2;  // Uniquely identifies the param at device scope
}
```

---

## 🎯 Goal / Outcome
GetParam can be queried through a GRPC endpoint
```
grpcurl -plaintext \
  -import-path ~/Catena/smpte/interface/proto \
  -proto service.proto \
  -d '{"slot": 0, "oid": "/counter"}' \
  localhost:6254 st2138.CatenaService/GetParam
```
```
{
  "oid": "counter",
  "param": {
    "name": {
      "displayStrings": {
        "en": "Counter",
        "es": "Contador",
        "fr": "Compteur",
        "ja": "カウンター"
      }
    },
    "type": "INT32",
    "precision": 2,
    "value": {
      "int32Value": 21
    },
    "minimalSet": true
  }
}
```

---

## ✅ Definition of Done (DoD)
A list of TODOs that need to be done for the issue to be done:
- [ ] Feature is implemented and works as expected
- [ ] Edge cases are handled
- [ ] Tests are added or updated (if applicable)
- [ ] Documentation is updated (if needed)
- [ ] Code is reviewed and approved

---

## 🛠️ Implementation Notes (Optional)

---

## 🔗 Related Issues / Links

---
## 🚧 Risks / Open Questions (Optional)

---

### PR DESCRIPTION
**Pull request by nelgd**
## 📖 Description

Implements the `GetParam` endpoint in the Go SDK for both the gRPC and REST transports. `GetParam` was previously a gRPC stub returning `Unimplemented` and had no REST route at all. It now returns the full parameter (metadata + current value) for a given slot/OID, wrapped as a `DeviceComponent.ComponentParam` (`{oid, param}`), following the same handler → runtime → transport pattern used by `GetValue`/`ParamInfoRequest`. A single `RegisterGetParamHandler` registration serves both transports.

---

## 🎯 Goal / Outcome

After merge, a client can call `st2138.CatenaService/GetParam` with `{slot, oid}`
and receive the complete `Param` (name, type, value, minimalSet, etc.). 

---

## ✅ Definition of Done (DoD)

- [x] gRPC `GetParam` method implemented (
- [x] REST route `GET /st2138-api/v1/{slot}/param/{fqoid...}` + `handleParamEndpoint` (GET-only, fqoid required, proto-JSON response, error mapping)
- [x] Test stub (`stubServerRuntime`) updated to satisfy the interface
- [x] gRPC tests: success (oid trimming + value/minimalSet), invalid slot, handler error; stale "unimplemented" cases removed
- [x] REST tests: success, nested fqoid, missing fqoid (400), method-not-allowed (405), handler error (404), plus transport-context propagation
- [x] Example handler registered in `oneOfEverything` (slot 0) so both documented calls work; endpoint guides updated for gRPC and REST
- [x] Docs updated (`GRPC.md`, `REST.md`)
- [x] `go build`, `go vet`, and `go test ./...` pass

---

## 🛠️ Implementation Notes (Optional)

- `GetParamHandler` returns `(*catena.Param, StatusResult)`. It does not use the
`Reply`/`ReplyError` generics because the `ResponseType` constraint is
`Value | Asset | Device`; it returns plain tuples, mirroring
`InvokeParamInfoHandler`.
- Both transports build the same `protos.DeviceComponent_ComponentParam{Oid, Param}`.
gRPC serializes it directly; REST serializes via `WriteProtoJSON` (proto field
names, unpopulated fields omitted), which yields the snake_case spec shape and
naturally omits `response:false` for params.
- A new `EndpointGetParam` enum value was inserted into `EndpointType` (after
`EndpointGetValue`), shifting subsequent iota values. This is internal-only
(used by the access-handler gate), not persisted, so the renumbering is safe.
- The REST handler trims a leading `/` from the response `oid` for parity with
gRPC, even though REST path-derived fqoids never carry one (harmless).

---

## 🔗 Related Issues / Links

Closes issue #1378 

Closes issue #1368

### CODE REVIEW

**Comment by nelgd** [comment]
Going to add in the REST endpoint on this pr. Makes more sense to have rest and grpc in one pr

File: sdks/go/pkg/catena/server.go [hunk 1]
```diff
@@ -122,6 +123,9 @@ func (ctx HandlerContext) HasAnyReadScope() bool {
 type DeviceHandler func(slot uint16, ctx HandlerContext) (Device, StatusResult)
 type GetValueHandler func(slot uint16, fqoid string, ctx HandlerContext) (Value, StatusResult)
 
+// GetParamHandler returns the full parameter (metadata + value) for a slot/fqoid.
+type GetParamHandler func(slot uint16, fqoid string, ctx HandlerContext) (*Param, StatusResult)
```
Comment (abrown-RV) [inline sdks/go/pkg/catena/server.go]:
```suggestion
type GetParamHandler func(slot uint16, fqoid string, ctx HandlerContext) (Param, StatusResult)
```
may as well stay consistent with the rest of the handlers. The wrapper has a ref to the proto, so copying the wrapper struct by value is just about the same as passing by ref.

File: sdks/go/pkg/catena/server.go
Comment (abrown-RV) [inline sdks/go/pkg/catena/server.go:1]: needs tests

File: sdks/go/examples/oneOfEverything/get_param_handler.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+// only) or ParamInfo (descriptor only).
+func registerGetParamHandlers(srv catena.Server, counter *CounterState, state *ExampleState) {
+	srv.RegisterGetParamHandler(0, func(slot uint16, fqoid string, ctx catena.HandlerContext) (*catena.Param, catena.StatusResult) {
+		logger.Info("GetParam", "slot", slot, "fqoid", fqoid)
+		if !ctx.HasReadScope(catena.ScopeCfg) {
+			return nil, catena.StatusWithCode(catena.StatusCodePermissionDenied, "configuration scope required")
+		}
+
+		switch fqoid {
+		case "counter", "/counter":
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/get_param_handler.go]: fqoids do not have leading slashes

File: sdks/go/examples/oneOfEverything/get_param_handler.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+			param := catena.NewParamInt32(counter.GetValue()).
+				WithName(catena.NewPolyglotText("en", "Counter").
+					With("es", "Contador").
+					With("fr", "Compteur").
+					With("ja", "カウンター")).
+				WithMinimalSet(true)
```
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/get_param_handler.go]: this should just match the device handler... maybe we can have a bunch of param maker helper funcs? that both device and get param handlers can call? param_makers.go or something?

File: sdks/go/pkg/transports/rest_transport.go [hunk 4]
```diff
@@ -589,6 +591,41 @@ func (t *RestTransport) handleAssetEndpoint(w http.ResponseWriter, r *http.Reque
...
+	if param == nil || param.Proto == nil {
+		t.writeHTTPStatusResult(w, catena.StatusWithCode(catena.StatusCodeInternal, "param returned nil"))
+		return
+	}
+
+	component := &protos.DeviceComponent_ComponentParam{
+		Oid:   strings.TrimPrefix(fqoid, "/"),
+		Param: param.Proto,
+	}
+	if err := WriteProtoJSON(w, component, http.StatusOK); err != nil {
```
Comment (abrown-RV) [inline sdks/go/pkg/transports/rest_transport.go]: this will need the same post-processing stuff to provide 0s. Constraints have a min_value field that is often 0 which must be included but the marshaller won't as we know and love.

also if the value is 0/0.0/""/etc. it will need to be added back in

(or do the emit defaults true thing, and strip out the extras, dealers choice XD)

**Review by abrown-RV — CHANGES_REQUESTED** [review]

File: sdks/go/examples/oneOfEverything/get_param_handler.go (hunk 3 above)
Comment (nelgd) [reply sdks/go/examples/oneOfEverything/get_param_handler.go]: Good idea, ill add that in

**Comment by nelgd** [comment]
bugbot run

File: sdks/go/pkg/transports/json_utils.go [hunk 5]
```diff
@@ -269,6 +286,50 @@ func cleanDeviceJSON(data []byte) ([]byte, error) {
...
+	// then removed, dropping the value the caller asked for.
+	var savedValue *fastjson.Value
+	if paramVal := v.Get("param"); paramVal != nil {
+		if val := paramVal.Get("value"); val != nil && val.Type() == fastjson.TypeObject {
+			savedValue = val
+			paramVal.Del("value")
+		}
+	}
+
+	deleteEmptyValues(v)
```
Comment (abrown-RV) [inline sdks/go/pkg/transports/json_utils.go]: This is exposing an issue with device request as well. It would have the same issue. The purpose is cleaning up things like `access_scope`, might be better to make deleteEmptyValues not touch strings and then deleteZeroFields or something like it can handle strings specifically targeted at a list

Also now that I'm looking at it later, `deleteEmptyValues` suggests its touching the `value` field, but it just means value from the json key/value pair sense.

File: sdks/go/examples/oneOfEverything/get_param_handler.go
Comment (abrown-RV) [inline sdks/go/examples/oneOfEverything/get_param_handler.go:1]: should do all slots as well, sorry I didn't make that clear

**Review by abrown-RV — CHANGES_REQUESTED** [review]

File: sdks/go/pkg/transports/grpc_transport.go [hunk 6]
```diff
@@ -430,10 +431,31 @@ func (s *catenaService) ExecuteCommand(req *protos.ExecuteCommandPayload, stream
...
+		return nil, status.Error(ToGRPCCode(result.Code), result.Error)
+	}
+	if param.Proto == nil {
+		logger.Error("GetParam returned nil param", "slot", slot, "fqoid", fqoid)
+		return nil, status.Error(codes.Internal, "param returned nil")
+	}
+
+	return &protos.DeviceComponent_ComponentParam{
+		Oid:   strings.TrimPrefix(fqoid, "/"),
+		Param: param.Proto,
```
[the remark this answers is not in the document]
Comment (abrown-RV) [reply sdks/go/pkg/transports/grpc_transport.go:456]: Fqoid do not start with leading slashes, if the endpoint is called with a leading slash, that is an error, not something we fix for them. There should be no trimming at all.

File: sdks/go/pkg/transports/rest_transport.go [hunk 7]
```diff
@@ -589,6 +591,52 @@ func (t *RestTransport) handleAssetEndpoint(w http.ResponseWriter, r *http.Reque
...
+	if r.Method != http.MethodGet {
+		t.writeHTTPMethodNotAllowed(w, "only GET allowed")
+		return
+	}
+
+	fqoid := strings.Join(pathParts, "/")
+	if fqoid == "" {
+		t.writeHTTPStatusResult(w, catena.StatusWithCode(catena.StatusCodeInvalidArgument, "request must include fqoid"))
+		return
+	}
```
[the remark this answers is not in the document]
Comment (nelgd) [reply sdks/go/pkg/transports/rest_transport.go:618]: Not sure if this is how it should be or if REST and gRPC should be handled the same way

[the remark this answers is not in the document]
Comment (abrown-RV) [reply sdks/go/pkg/transports/rest_transport.go:618]: make a ticket, we'll revist

File: sdks/go/pkg/transports/json_utils.go [hunk 8]
```diff
@@ -377,7 +440,7 @@ func deleteEmptyValues(v *fastjson.Value) bool {
 
 	case fastjson.TypeArray:
 		for _, elem := range v.GetArray() {
-			deleteEmptyValues(elem)
+			deleteEmptyFields(elem)
```
[the remark this answers is not in the document]
Comment (abrown-RV) [reply sdks/go/pkg/transports/json_utils.go:440]: The example cited by bugbot here is not real. But I guess it doesn't really matter if the removal is positive or negative. So its all good.

File: sdks/go/pkg/transports/rest_transport.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (nelgd) [reply sdks/go/pkg/transports/rest_transport.go:618]: #1437

**Review by abrown-RV — APPROVED** [review]
